## 2.1 函数与程序设计

摘自：《程序设计导论，Python语言实践》p137

函数的重要意义在于，可以在程序中清晰地分离不同的任务，而且为代码复用，提供了一个通用的机制。

### 2.1.1 一个原则：重要的程序设计理念

**在计算任务中，任何时候只要可以清晰地分离任务，则建议使用函数分离任务。**

首先解释下，什么是调和数：

![](https://img.imotao.com/i/2024/02/06/65c1f0855f112.png)

下面看一个调和数函数 harmonics(n) 的例子：

`print("调和数 H({}) = {}".format(n, harmonic_number(n)))` 

解释：这里使用了 Python 的字符串格式化方法 .format() 来插入变量 n 和 harmonic_number(n) 的值到字符串中。


In [8]:
def harmonic_number(n):
    """
    计算调和数H(n)
    """
    if n < 1:
        return "n必须大于等于1"
    result = 0.0
    for i in range(1, n + 1):
        result += 1 / i
    return result

# 测试
n = int(input("请输入一个整数 n："))
print("调和数 H({}) = {}".format(n, harmonic_number(n)))


调和数 H(100000000) = 18.997896413852555


下面再看一个书中写的调和数函数：

![](https://img.imotao.com/i/2024/02/06/65c1f69e3710e.png)

上图为测试输出 ⬆



In [ ]:
#-----------------------------------------------------------------------
# harmonicf.py
# https://introcs.cs.princeton.edu/python/21function/harmonicf.py.html
#-----------------------------------------------------------------------

import stdio
import sys

# Return the nth harmonic number.
def harmonic(n): # n代表调和数的阶数
    total = 0.0

    # 这就是 Python 中实现调和数的函数的主要代码
    for i in range(1, n+1):
        # total += 1.0 / float(i)
        total = total + 1.0 / float(i)
    return total

# Write to standard output the harmonic numbers specified as
# command-line arguments.

for j in range(1, len(sys.argv)):

    arg = int(sys.argv[j]) #获取命令行参数，并进行显式类型转换
    value = harmonic(arg) # arg = n，调用调和数函数，返回值 total，再赋值给 value
    stdio.writeln(value) 


### 2.1.2 程序追踪：函数的调用和返回值跟踪

跟踪函数调用控制流程的一个最简单方法是，假定每个程序在调用时输出自己的函数名和各参数值，并在返回时输出返回值，调用函数后采用缩进显示输出格式，返回值的输出格式则取消缩进。


In [10]:
def harmonic_number(n):
    """
    计算调和数H(n)
    """
    if n < 1:
        return "n必须大于等于1"
    result = 0.0
    for i in range(1, n + 1):
        result += 1 / i
    return result

# 测试
n = int(input("请输入一个整数 n："))
print("调和数 H({}) = {}".format(n, harmonic_number(n)))


调和数 H(2) = 1.5


i=1
arg=1
harmonic(1)
    total=0.0
    total=1.0
    return 1.0
value=1.0

i=2
arg=2
harmonic(2)
    total=0.0
    total=1.0
    total=1.5
    return 1.5
value=1.5

### 2.1.3 函数概念与Python语法构造

摘自《程序设计导论，Python语言实践》p141

| 函数概念        | Python 语法构造 | 功能描述                                      |
|----------------|-----------------|----------------------------------------------|
| 函数           | 函数             | 映射                                    |
| 输入值           | 实际参数             | 函数的输入值                                  |
| 输出值       | 返回值         | 函数的输出                                  |
| 公式         | 函数体           | 函数定义                                  |
| 独立变量       | 形式参数变量     | 输入值的符号占位符                            |

以上是表格，用于描述函数概念和 Python 语言语法构造之间的对应关系。

## 2.2 参数传递与返回值机制

详见 `Dept2/01-Function.ipynb` 。
